src/TuxState.js

```diff
@@ -0,0 +1,650 @@
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToSplice, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that current state at this key is an array
+      invariantArrayCheck(currentStateAtKey);
+      //check that newPropsAtKey is an array of arguments to be passed into the splice invocation
+      invariantArrayCheck(newPropsAtKey);
+      //invoke splice on currentStateAtKey with the array of arguments in newPropsAtKey
+      Array.prototype.splice.apply(currentStateAtKey, newPropsAtKey);
+      newState[key] = currentStateAtKey;
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //concatToEndOfState FUNCTION: concat the arrays of values in the deepest keys on the passed in object to the end of the corresponding arrays in the deepest keys in the current state. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToShift OBJECT: required object argument with deepest keys being a boolean
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  concatToEndOfState: function(propsToConcat, callback) {
+    //throw error if propsToConcat is undefined or not an object
+    invariantArgCheck(propsToConcat);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToConcat, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that currentStateAtKey is an array
+      invariantArrayCheck(currentStateAtKey);
+      //check that newProps at this key is an array
+      invariantArrayCheck(newPropsAtKey);
+      //concat newPropsAtKey to the end of currentStateAtKey
+      newState[key] = currentStateAtKey.concat(newPropsAtKey);
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //concatToFrontOfState FUNCTION: concat the arrays of values in the deepest keys on the passed in object to the front of the corresponding arrays in the deepest keys in the current state. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToConcat OBJECT: required object argument with deepest keys being a boolean
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  concatToFrontOfState: function(propsToConcat, callback) {
+    //throw error if propsToConcat is undefined or not an object
+    invariantArgCheck(propsToConcat);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
```

Comment:
I don't think we need to invoke `assign` here because we are already making a shallow copy of state at the first line in `buildNewState`.